File: solutions/211.design-add-and-search-words-data-structure.py
#
# @lc app=leetcode id=211 lang=python3
#
# [211] Design Add and Search Words Data Structure
#

# @lc code=start
class Node:
    def __init__(self):
        self.children = [None] * 26
        self.is_word_end = False
        self.word_max_length = 0

class WordDictionary:

    def __init__(self):
        self.root = Node()

    def addWord(self, word: str) -> None:
        node = self.root

        for c in word:
            index = ord(c) - ord('a')
            if not node.children[index]:
                node.children[index] = Node()
            node = node.children[index]

        node.is_word_end = True
        self.root.word_max_length = max(self.root.word_max_length, len(word))

    def search(self, word: str) -> bool:
        node_length = self.root
        if not node_length or node_length.word_max_length < len(word):
            return False

        def dfs(word, position, node):
            if position == len(word):
                return node.is_word_end

            current_char = word[position]
            if current_char != '.':
                next_node = node.children[ord(current_char) - ord('a')]
                return next_node and dfs(word, position + 1, next_node)

            # Process all possible chars when there is '.' in the word
            for i in range(26):
                next_node = node.children[i]
                if next_node and dfs(word, position + 1, next_node):
                    return True

            return False

        return dfs(word, 0, self.root)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
# @lc code=end

File: solutions/211.design-add-and-search-words-data-structure.java
/*
 * @lc app=leetcode id=211 lang=java
 *
 * [211] Design Add and Search Words Data Structure
 */

// @lc code=start
class WordDictionary {

    Node root;
    class Node{
        Node[] children = new Node[26];
        boolean isEndOfWord = false;
        int maxLettersWord = 0;     // the longest word length. Set for first letter in first node after root.
    }

    public WordDictionary() {
        root = new Node();
    }

    public void addWord(String word) {
        Node node = root;
        for(char c : word.toCharArray()) {
            int index = c - 'a';
            if (node.children[index] == null) {
                node.children[index] = new Node();
            }
            node = node.children[index];
        }
        node.isEndOfWord = true;
        int index = word.charAt(0) - 'a';
        root.children[index].maxLettersWord = Math.max(root.children[index].maxLettersWord, word.length());
        root.maxLettersWord = Math.max(root.maxLettersWord, word.length());
    }

    public boolean search(String word) {
        Node firstNode =  word.charAt(0) == '.' ? root : root.children[word.charAt(0) - 'a'];
        if (firstNode == null || firstNode.maxLettersWord < word.length()) {
            return false;
        }

        return dfs(word, 0, root);
    }

    public boolean dfs(String word, int position, Node node) {
        if (word.length() == position) {
            return node.isEndOfWord;
        }

        char currentChar = word.charAt(position);
        if (currentChar != '.') {
            Node nextNode = node.children[currentChar - 'a'];
            return nextNode != null && dfs(word, position + 1, nextNode);
        }

        // Process all possible chars when there is '.' in the word
        for (int i = 0; i < 26; i++) {
            Node nextNode = node.children[i];
            if (nextNode != null && dfs(word, position + 1, nextNode)) {
                return true;
            }
        }

        return false;
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */
// @lc code=end

